OpenTelemetry instrumentation example (#296)
temporalio/samples-go: otel/README.md

Reviewer: Can we call this folder `opentelemetry`?

Author: yes, I was thinking about that.

### Steps to run this sample:
1) Run a [Temporal service](https://github.com/temporalio/samples-go/tree/main/#how-to-use).

> One way could be just to use the temporal CLI.  

Reviewer: ```suggestion
One way could be just to use the Temporal CLI.  
```

```bash

> temporal server start-dev --ui-port 8089

Reviewer: ```suggestion
temporal server start-dev
```

```

2) Run the following command to start the worker
```bash
go run otel/worker/main.go
```

> 3) Run the following command to start the example

Reviewer: ```suggestion
3) In another terminal, run the following command to run the workflow
```

```bash
go run otel/starter/main.go
```

The example outputs the traces in the stdout, both the worker and the starter.  

> [Trace output](./trace.json) collects the spans generated. Please note that in order to have the valid json they have been put in a JSON array but they are returned as single JSON objects.  

Reviewer: I don't think we need to put the pre-generated trace JSON in the repo

Author: I was thinking this could help to see what to expect when running the example, if those don't show up when running the example then there's something missing.

Reviewer: The image makes sense there, but the full JSON trace I don't think has that much value. I think it's ok to expect people to run the sample if they want to see the big JSON output.

If all is needed is to see Workflows and Activities there's no need to set up instrumentation for the Temporal cluster.  

In order to send the traces to a real service you need to replace

```go
exp, err := stdouttrace.New(stdouttrace.WithPrettyPrint())
if err != nil {
    log.Fatalln("failed to initialize stdouttrace exporter", err)
}
```
with  
```go
// Configure a new OTLP exporter using environment variables for sending data to Honeycomb over gRPC
clientOTel := otlptracegrpc.NewClient()
exp, err := otlptrace.New(ctx, clientOTel)
if err != nil {
    log.Fatalf("failed to initialize exporter: %e", err)
}
```

And provide the required additional parameters like the OTLP endpoint.  
For many services that would mean just to set the standard OTeL env vars like:

```
OTEL_SERVICE_NAME
OTEL_EXPORTER_OTLP_ENDPOINT
OTEL_EXPORTER_OTLP_HEADERS
```

As an example this is what is the rendered by Honeycomb.io.  

> <img width="970" alt="Screenshot 2023-07-13 at 21 44 24" src="https://github.com/emanuelef/samples-go/assets/48717/d6e7fa3b-3604-4344-8e61-a81e0a02acd2">

Reviewer: Would rather not have a link to your fork. Feel free to include the image as part of the sample like we do in https://github.com/temporalio/samples-go/tree/main/synchronous-proxy

Author: Added the image but doesn't seem to be good practice as it will increase the size of the repo when cloning.
